### PR TITLE
feat: implement xdg-dialog-v1, modal dialog handling and fix issues with overlapping children windows

### LINKED ISSUE #2776: xdg-dialog-v1
From the Poland summit

* Position on top of the window
* Movable
* Floating
* Don’t dim window beneath

For COSMIC apps
  * Use this protocol
  * Close when un-focused

### PR DESCRIPTION
**Pull request by mayakwd**
Implements [xdg-dialog-v1](https://wayland.app/protocols/xdg-dialog-v1) and makes modal dialogs behave consistently for both Wayland and X11 clients. Also, it fixes issues with parent windows overlapping transient children (huge pain, simple solution) 

Fixes:
- #2446 (the modal-dialog behavior part: focus, stacking, and dimming for a window blocked by a modal child).
- Transient children overlapped by parent window issues fixes: 
  - #1805,
  - #1989,
  - #2259, 
  - pop-os/cosmic-epoch#1944, 
  - pop-os/cosmic-epoch#3171
  - probably more (some of them partially)

https://github.com/user-attachments/assets/55db8d6a-0404-40a6-a8ed-82d278b79379

## Important scope note
#1989's thread mixes modal and *non-modal* transient windows, which end up behind their parent toplevel. For example:
- gedit's "About" window: GTK3 apps on Wayland can't signal modality (only via gtk-shell protocol, which is not supported). The GTK4 gedit build displays the "About" window as a modal one.
- JetBrains products are now on Wayland, and they implemented their own client-side focus-managing system, intentionally moving focus to pseudo-modal windows like Settings)

Non-modal transient windows ending up behind their parent is a separate mechanism from modality — and this PR addresses it too: the parent-child check is unified in CosmicSurface::is_parent_of (xdg parent for Wayland, WM_TRANSIENT_FOR for X11), and raise_with_children now uses it. Previously, only Wayland children were raised with their parents, which is why mostly X11 apps (Reaper, Bitwig, JetBrains) were hit the hardest.

## Note on smithay bump
The X11 side relies on the smithay change I landed in Smithay/smithay#2062 (`_NET_WM_STATE_MODAL` client messages -> `XwmHandler::modal_request` / `unmodal_request`, `X11Surface::set_modal` / `is_modal`). **UPD:** no smithay bump needed anymore since cosmic-comp's master already includes everything needed.

## What it does

A toplevel that has a visible modal child dialog is now treated as *blocked*:

- **Focus**
  Keyboard focus on the blocked parent is redirected to the modal dialog, so the parent can't regain focus while the dialog is open. The redirect lives in a single choke point in `Shell::set_focus`, so every focus path is covered. Closing the dialog switches focus back to its parent with an intentional focus override to provide a good UX and avoid focusing on an unrelated window in the focus stack.
- **Stacking**
  The dialog always stays above its parent. Focusing a modal dialog raises its entire parent chain, so the parent can't end up buried behind unrelated windows.
- **Refused actions**
  Move/resize/maximize/minimize/fullscreen of a blocked parent is refused, and the dialog shortly shakes to point the user to it. Clicking the blocked parent also shakes it (watch the video).
- ~~**Dimming**
   The blocked parent's content is dimmed with a short fade.~~
- **Transient stacking (modal and non-modal)**
  Transient children are kept above their parents for both Wayland and X11. The parent-child relation is unified in `is_parent_of`. Previously, the `raise_with_children` logic only recognized Wayland parents, so X11 transient children could end up buried under their parent.
- **Placement**
   Modal dialog opens centered on top of its parent. This is initial placement only. The dialog stays floating and movable.
  

## What I've tested
- Wayland apps.
- X11: a GTK app under `GDK_BACKEND=x11`, including toggling `set_modal(true)` *after* the window is mapped (checking the new smithay callback from Smithay/smithay#2062).
- Some applications from the mentioned issues: Reaper, Bitwig, Godot
- Scenarios
   - focus redirect
   - shake on each refused action and on click
   - ~~dim fade in/out~~
   - dialog close -> focus returns to parent
   - modal-of-modal chains
   - tiled/floating/stacking

## What I didn't do
There are pre-existing tiling/stacking/transient gaps that the modal support just makes more visible. I left them out on purpose - they touch a lot of things, and it's better to discuss things and land them separately:

- A transient family isn't moved between workspaces as a whole, so a modal dialog can be sent to a different workspace from its parent. In that case, focus restoration then falls back to the focus stack, not a parent in a different space.
- A dialog of an inactive stack tab stays visible over the active tab.
- ~~A fullscreen parent isn't dimmed.~~

*Not relevant for this PR but still valuable:*
- While building the dim, I found an issue related to geometry presented by `CosmicMapped::active_window_geometry()`. For a non-stacked X11 window, its `Window` branch counts the SSD header twice (the `loc` already skips the header, and the `size` adds `SSD_HEIGHT` again), while the `Stack` branch is content-only. It also affects the screencopy cursor hit-test and the tiling group overlay. For now, I've made a workaround (with an explanation in the code comments and `TODO` markers). 

- [x] I have disclosed use of any AI generated code in my commit messages.
  - If you are using an LLM, and do not fully understand the changes it is making to the code base, do not create a PR.
  - In our experience, AI generated code often results in overly complex code that lacks enough context for a proper fix or feature inclusion. This results in considerably longer code reviews. Due to this, AI authored or partially authored PRs may be closed without comment.
- [x] I understand these changes in full and will be able to respond to review comments.
- [x] My contribution is tested and working as described.
- [x] My change is accurately described in the commit message.
- [x] I have read the [Developer Certificate of Origin](https://developercertificate.org/) and certify my contribution under its conditions.

### CODE REVIEW

**Comment by dkonigsberg** [comment]
Just one question about this:
> gedit's "About" window is not modal (it shouldn't block gedit's main window)

gedit's "About" window absolutely does block gedit's main window on Gnome and KDE, so how exactly is it not modal?  (Unless there's another definition of window state that looks modal to a user, but is technically something else.)

Also, I have to thank you immensely for taking on this task.

**Comment by mayakwd** [comment]
> gedit's "About" window absolutely does block gedit's main window on Gnome and KDE, so how exactly is it not modal? (Unless there's another definition of window state that looks modal to a user, but is technically something else.)

Hey @dkonigsberg, thanks for a great question! 

It made me think that I haven't done deep enough research. :-) Yet I've checked gedit once again. So your question is totally valid. 

gedit's "About" window is modal by design, but the installation I have and tested uses GTK3. GTK3 does not support a modality over the Wayland protocol, only X11. Modality support for Wayland was introduced in GTK4. 

GTK uses the `GDK_BACKEND` environment variable to determine which backend to use.  Since COSMIC exports `GDK_BACKEND=wayland,x11` - `wayland` always wins by priority. 

If we enforce X11 usage `GDK_BACKEND=x11 gedit`, then we'll see that the "About" window is modal and blocking the parent, as expected :-)

**Comment by Drakulix** [comment]
> gedit's "About" window is modal by design, but the installation I have and tested uses GTK3. GTK3 does not support a modality over the Wayland protocol, only X11. Modality support for Wayland was introduced in GTK4.

Nitpick just for completeness sake: It does support modality, but only via the private gtk-shell protocol: https://wayland.app/protocols/gtk-shell#gtk_surface1:request:set_modal

Which we will not implement for obvious reasons.

**Comment by HigherOrderLogic** [comment]
Any chance this can be partially extracted to Smithay? I'm also looking into implement this protocol for Niri.

**Comment by mayakwd** [comment]
@HigherOrderLogic, surprisingly, there is nothing to extract to Smithay. Smithay already has everything required for implementation - it already provides [XdgDialogState](https://docs.rs/smithay/latest/smithay/wayland/shell/xdg/dialog/struct.XdgDialogState.html),  https://github.com/Smithay/smithay/pull/2062, added modality requests handling.

The entire implementation in this PR only implements the modality in terms of cosmic-comp window handling and doesn't include anything to extract to Smithay.

**Comment by HigherOrderLogic** [comment]
> surprisingly, there is nothing to extract to Smithay

Thanks. Seems like the infra have been merged into Smithay some time ago, but I just missed it.

And by the way, the app you showcased was Gnome Calculator right? Did you do any special setting to the Calculator app's preference dialog to get it as a separate window? Somehow mine still show up as embedded into the main window instead.

**Comment by mayakwd** [comment]
> The app you showcased was Gnome Calculator right? Did you do any special setting to the Calculator app's preference dialog to get it as a separate window?

It is Gnome Calculator (pre-installed), I didn't do anything special.

**Comment by mayakwd** [comment]
Dropped smithay bump commit, rebased on master.
Slightly updated description.

**Comment by mayakwd** [comment]
@Drakulix, hi! 
Does it make sense to update this PR to resolve the conflicts, or should I just close it?

**Comment by Drakulix** [comment]
> @Drakulix, hi! Does it make sense to update this PR to resolve the conflicts, or should I just close it?

Hi! Sorry, we certainly want this feature, however I still need to discuss exact behavior with our UX team (for things like dimming the window).

So please don't close it, but maybe just wait for me to come back to this in a couple of weeks? (Things are a little busy atm.)

**Comment by mayakwd** [comment]
> Hi! Sorry, we certainly want this feature, however I still need to discuss exact behavior with our UX team (for things like dimming the window).
> 
> So please don't close it, but maybe just wait for me to come back to this in a couple of weeks? (Things are a little busy atm.)

Of course! I apologize if my message gave the impression that I was rushing you. My original intention was to find out whether it makes sense to support this PR, whether a more suitable alternative is in preparation or already available, or maybe it's worth splitting this PR in some way.

Thank you, I'll just update it then to align with the new changes.

**Comment by WatchMkr** [comment]
The behavior we decided on at the Poland summit:

- Position on top of the window
- Movable
- Floating
- Don’t dim window beneath

**Comment by mayakwd** [comment]
@WatchMkr,

Ah, sad that I failed to join you in Warsaw.
So it is movable and floating (I hope I understand it correctly), and I have to remove the dimmer. 

What about shaking? Is it relevant, or should I remove it as well?

**Comment by WatchMkr** [comment]
Sorry we missed you!

The shake is good to bring the users attention to the blocking modal. Since it's already movable and floating, I think it's just removing dimming and positioning the model on top of its parent.

Thanks for working on this!

**Comment by mayakwd** [comment]
> Sorry we missed you!
> 
> The shake is good to bring the users attention to the blocking modal. Since it's already movable and floating, I think it's just removing dimming and positioning the model on top of its parent.
> 
> Thanks for working on this!

Thanks! Updated PR to match the summit decision:
- dimming is gone
- modal now opens centered on top of parent
- it was already floating and movable

Also, rebased on top of master, so conflicts were resolved (CI failures are inherited from master).

File: src/shell/layout/floating/mod.rs [hunk 1]
```diff
@@ -406,6 +434,16 @@ impl FloatingLayout {
...
 
+    /// Geometry of the window `mapped` is transient for, if that parent is mapped here.
+    fn parent_geometry(&self, mapped: &CosmicMapped) -> Option<Rectangle<i32, Local>> {
+        let child = mapped.active_window();
+        self.space
+            .elements()
+            .find(|elem| elem.active_window().is_parent_of(&child))
+            .and_then(|parent| self.space.element_geometry(parent))
+            .map(RectExt::as_local)
+    }
```
Comment (Drakulix) [inline src/shell/layout/floating/mod.rs:445]: So modals will only be centered for floating and sticky windows? What about a model for a tiled window?

**Review by Drakulix — COMMENTED** [review]
Looks great code wise, just a single question.

File: src/shell/layout/floating/mod.rs (hunk 1 above)
Comment (mayakwd) [reply src/shell/layout/floating/mod.rs:445]: Uh, that's right, I just forgot about tiled windows. Working on that. I believe that we need a special branch for sticky windows as well.

Comment (mayakwd) [reply src/shell/layout/floating/mod.rs:445]: Sorry for the delay, during more thorough testing I found several edge cases I'd like to fix, before bringing complete solution to the project. I hope to finish this week (busy days).

Comment (mayakwd) [reply src/shell/layout/floating/mod.rs:449]: Update - fixed several edge-cases and rewrote some paths (PR description also updated).
What changed:

- Dialog used to open on whatever output was active, so on a multi-monitor setup it regularly appeared away from the window it blocks. It now opens on its parent's output and workspace, with modal dialogs additionally centered on the parent. There is more info in PR description (Notes on placement design)
- Locating the parent and its rectangle goes through one unified path now, so centering is correct for a floating, tiled, stacked, sticky or fullscreen parent, on any output or workspace. A fullscreen parent previously yielded no anchor at all and the dialog fell back to cascade placement.
- Dialog's own child window was buried when the dialog was clicked, because the focused window was raised above its own descendants. The raise walk also restacked layers that didn't hold the focused window at all.
- The window menu's Resize entries bypassed the modal block that every other path honored.
- Shake now is  skipped for maximized/fullscreen dialogs.
- `WM_TRANSIENT_FOR` isn't validated by the server, so walk could hang on self-transient windows or cyclic chains.
